internal/vcs/git/commits.go

```
@@ -336,6 +336,99 @@ type onelineCommit struct {
+// batch up calls to scan.
+func logOnelineBatchScanner(scan func() (*onelineCommit, error), maxBatchSize int, debounce time.Duration) (next func() ([]*onelineCommit, error), cleanup func()) {
+	// done is closed to indicate cleaning up
+	done := make(chan struct{})
+	cleanup = func() {
+		close(done)
+	}
+
+	// We can't call scan with a timeout. So we start a goroutine which sends
+	// the output of scan down a channel. When done is closed, this goroutine
+	// will stop running.
+	type result struct {
+		commit *onelineCommit
+		err    error
+	}
+	// have a cap of maxBatchSize so we potentially have maxBatchSize ready to
+	// send.
+	resultC := make(chan result, maxBatchSize)
+	go func() {
+		for {
+			commit, err := scan()
+			select {
+			case resultC <- result{commit: commit, err: err}:
+			case <-done:
+				return
+			}
+			if err != nil {
+				return
+			}
+		}
+	}()
+
+	var err error
+	next = func() ([]*onelineCommit, error) {
+		// The previous call may have had an error but also had commits to
+		// return. In that case we didn't return the error, so need to return
+		// it now.
+		if err != nil {
+			return nil, err
+		}
+
+		// Stop scanning if we haven't found maxBatchSize by deadline.
+		deadline := time.NewTimer(debounce)
+		defer deadline.Stop()
+
+		// We always want 1 result. Ignore deadline for the first result.
+		r := <-resultC
+		if r.err != nil {
+			err = r.err
+			return nil, r.err
+		}
+
+		// Add to commits until we hit maxBatchSize or we are passed
+		// deadline. If we encounter an error set err.
+		commits := []*onelineCommit{r.commit}
+		timedout := false
+		for err == nil && len(commits) < maxBatchSize && !timedout {
+			select {
+			case r := <-resultC:
+				if r.err != nil {
+					err = r.err
+				} else {
+					commits = append(commits, r.commit)
+				}
+			case <-deadline.C:
+				timedout = true
+			}
+		}
+
+		if len(commits) > 0 {
+			// Note: err can be non-nil here. The next call to this
```

I am not sure I understand. We ignore errors if `len(commits)>0` (ok), but how does the next call to `next` return the error we ignored here?

---

err is outside of the scope of the function. If you look at the top of the function we return, we check if `err != nil` and if so return it then.

---

I've somewhat cleaned up this bit of code to remove the conditional. We are gaurenteed that len(commits) > 0. The comment has stayed the same though.